url416ee5b4ee_martinmarinov_HonProj_tgz-pJ8-martin_gui_MainJ8/src/martin/gui/quantum/Corrector.java

```diff
@@ -0,0 +1,391 @@
+			}
+		}
+		
+		g.setStroke(def);
+		
+		final Font defaultFont = g.getFont();
+		if (font == null) font = new Font(defaultFont.getFamily(), Font.PLAIN, DEFAULT_FONT_SIZE);
+		g.setFont(font);
+		final String sid = String.valueOf(t);
+		final FontMetrics fm = g.getFontMetrics();
+		final Rectangle2D f = fm.getStringBounds(sid, g);
+		int stx = (int) (p.x - f.getWidth()  / 2);
+		int sty = (int) (p.y - f.getHeight() / 2  + fm.getAscent());
+		g.drawString(sid, stx, sty);
+		g.setFont(defaultFont);
+		
+
+	}
+	
+	private CubicCurve2D.Double getBezier(final Point2D.Float p, final int sx, final int sy, final int ex, final int ey) {
+		
+		final int ddx = sx - ex;
+		final int ddy = sy - ey;
+		final double nang = Math.atan2(ddy, ddx)+Math.PI/2;
+		final int sin = (int) (Math.sin(nang) * arcd);
+		final int cos = (int) (Math.cos(nang) * arcd);
+		
+		final int cp1x = sx + cos;
+		final int cp1y = sy + sin;
+		final int cp2x = ex + cos;
+		final int cp2y = ey + sin;		
+		
+		if (p != null) {
+			p.x = (cp1x + cp2x) / 2;
+			p.y = (cp1y + cp2y) / 2;
+		}
+		
+		return new CubicCurve2D.Double(sx, sy, cp1x, cp1y, cp2x, cp2y, ex, ey);	
+	}
+	
+	private ArrayList<double[]> generatePoints(CubicCurve2D.Double bezier) {
+		final FlatteningPathIterator pi = new FlatteningPathIterator(bezier.getPathIterator(null), 0.1);
+
+		final ArrayList<double[]> path = new ArrayList<double[]>();
+
+		while (!pi.isDone()) {  
+
+			final double[] coordinates = new double[6];
+			pi.currentSegment(coordinates);
+
+			final double dx1 = coordinates[0] - i1.x;
+			final double dy1 = coordinates[1] - i1.y;
+			final double d1 = dx1 * dx1 + dy1 * dy1;
+			
+			if (d1 > QUBIT_SIZE * QUBIT_SIZE) {
+				final double dx2 = coordinates[0] - i2.x;
+				final double dy2 = coordinates[1] - i2.y;
+				final double d2 = dx2 * dx2 + dy2 * dy2;
+				
+				if (d2 > QUBIT_SIZE * QUBIT_SIZE)
+					path.add(coordinates);
+			}
+
+			pi.next();
+		}
+		
+		return path;
+	}
+
+	@Override
+	boolean isMouseOntop(int x, int y) {
+		final Point2D.Float p = new Point2D.Float();
+		final ArrayList<double[]> bpts = generatePoints(getBezier(null, i1.x,
+				i1.y, i2.x, i2.y));
+		final int sizeb = bpts.size();
+
+		if (sizeb > 1) {
+			final double[] first = bpts.get(0);
+			final double[] last = bpts.get(sizeb - 1);
+			final ArrayList<double[]> path = generatePoints(getBezier(p,
+					(int) first[0], (int) first[1], (int) last[0],
+					(int) last[1]));
+
+			final int size = path.size();
+
+			double[] prev = path.get(0);
+			for (int i = 1; i < size; i++) {
+				final double[] curr = path.get(i);
+				final double dx = x - curr[0];
+				final double dy = y - curr[1];
+				if (dx * dx + dy * dy < PICKING_DISTANCE * PICKING_DISTANCE)
+					return true;
+				if (distFromLine(x, y, prev[0], prev[1], curr[0], curr[1]) < PICKING_DISTANCE)
+					return true;
+				prev = curr;
+			}
+		}
+		return false;
+	}
+	
+	public double distFromLine(double px, double py, double sx, double sy, double ex, double ey) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.